docs/feature_rawhid.md

```diff
@@ -0,0 +1,119 @@
+# Raw HID
+
+Raw HID allows for bidirectional communication between keyboard and a host computer over HID interface. It allows for 32 byte bidirectional packet transfer in QMK.
+
+There are two main components to getting raw HID working with your keyboard.
+
+## Keyboard firmware
+
+The implementation is fairly straightforward for the firmware.
+In your `rules.mk` add:
+
+```make
+RAW_ENABLE = yes
+```
+
+In your `keymap.c` include `"raw_hid.h"` and implement the following:
+
+```C
+void raw_hid_receive(uint8_t *data, uint8_t length) {
+    // Your code goes here. data is the packet received from host.
+}
+```
+
+The `"raw_hid.h"` header also includes `void raw_hid_send(uint8_t *data, uint8_t length);`
+which allows sending packets from keyboard to host. It can be used for debugging when building your host by returning all data back to host to test communication.
+
+```C
+void raw_hid_receive(uint8_t *data, uint8_t length) {
+    raw_hid_send(data, length);
+}
+```
+
+All Raw HID packets by default have size `RAW_EPSIZE` which is `32 bytes`, the actual length is determined by `uint8_t length` both when sending and receiving.
+
+Make sure to flash raw enabled firmware before proceeding with working on the host side.
+
+## Host (PC/Mac/Linux)
+
+This is the more complicated part as it will require some digging.
+
+To connect your host computer to your keyboard with raw HID you need four pieces of information about your keyboard:
+
+1. Vendor ID
+2. Product ID
+3. Usage Page
+4. Usage
+
+The first two can easily be found in your keyboard's `config.h` in the keyboard's main directory under `VENDOR_ID` and `PRODUCT_ID`.
+
+The second two are slightly harder and can vary by vendor. Most of the time though for most QMK keyboards **usage page is `0xFF60`** and **usage is `0x0061`**.
+
+#### The harder way
+
+If these values don't work for you, you can easily discover the usage page and usage for your keyboard by finding the values for `tag 4` for usage page and `tag 8` for usage in the hid interface descriptor for your keyboard by using any form of hid dump or any hid library to find the descriptor and extract the values from the tags.
```

Comment:
This information is mostly unnecessary. The usage page and usage must *always* be `0xFF60`/`0x61`.

Reply:
This doesn't happen to be the case for the HID interface for my board, md ctrl. Idk about raw hid because it's not implemented but for opening an HID interface I had to use `0xFF31` and `0x0074`, though I don't know if that will be the case for raw hid when it's implemented. But you know better, if it isn't needed let's remove it, just confirm that's the case.

Reply:
Raw HID and Console are separate features. These are the descriptors for LUFA and ChibiOS:
https://github.com/qmk/qmk_firmware/blob/3eb82e047079230e55d3aeff901c4d0dce69fb3a/tmk_core/protocol/usb_descriptor.c#L233-L279
And for ATSAM (the formatting is a bit wonky, and the comments are a bit wrong):
https://github.com/qmk/qmk_firmware/blob/3eb82e047079230e55d3aeff901c4d0dce69fb3a/tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c#L638-L653
https://github.com/qmk/qmk_firmware/blob/3eb82e047079230e55d3aeff901c4d0dce69fb3a/tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c#L755-L772

Reply:
I know they're separate. I tested the interfaces in the descriptor and they came out wrong, regardless of the wrong comments, it just doesn't connect at all for any of the proper values according to the array index, they're not valid HID interfaces.

Reply:
Fixed. Still need to decide on which host implementation to present if any at all.

Reply:
For all intents and purposes there is no ATSAM implementation right now - the descriptor is there and will be sent to the host when it requests it, but the functions in rawhid.h are never called.